### PR TITLE
Add procedures for beta and final releases

### PR DESCRIPTION
**Pull request by tk0miya**
This describes release procedure for #4398.

### CODE REVIEW

File: utils/release-checklist
Comment (stephenfin) [inline utils/release-checklist:31]: We could fix this by making `master` a protected branch in GitHub settings. There is documentation [here](https://help.github.com/articles/about-protected-branches/) about that. Could we do that first?

File: utils/release-checklist
Comment (stephenfin) [inline utils/release-checklist]: This isn't valid rST. Also, you should use `git fetch` first

**Review by stephenfin — CHANGES_REQUESTED** [review]

File: utils/release-checklist
Comment (tk0miya) [reply utils/release-checklist:31]: Ah, I know that and also understand its goodness. On the other hand, I usually update CHANGES file and push it master branch directly. So the restriction is a bit boring to me...

File: utils/release-checklist
Comment (tk0miya) [reply utils/release-checklist]: Where is invalid?

I fixed about `git fetch`.

Comment (stephenfin) [reply utils/release-checklist]: Well, not invalid. It's just that the default role has no significance here. That's noted in the docs:

    The default role (`content`) has no special meaning by default. You are free to use it for
    anything you like, e.g. variable names; use the default_role config value to set it to a known
    role.

I think it would be better to format it as a literal with double backticks.

File: utils/release-checklist
Comment (stephenfin) [reply utils/release-checklist:31]: Well then that sounds like an issue we should resolve. It shouldn't be so hard to modify CHANGES. Could you add a TODO to change this policy once we figure out a better tool to help us here?

File: utils/release-checklist
Comment (tk0miya) [reply utils/release-checklist]: You're right. fixed.

File: utils/release-checklist
Comment (tk0miya) [reply utils/release-checklist:31]: No, this is release manual. It's not good to note a TODO.
Please file a new issue please.

Comment (stephenfin) [reply utils/release-checklist:31]: It is good. If we put it here, all maintainers know that we consider this part of the process to be suboptimal and that we want to improve it. If we put it in the bug tracker, it will be forgotten about. This is not a bug or RFE for Sphinx - it's a process improvement we'd like to see.

There are two TODOs.

- Find a way to ensure CHANGES updated with every pull request. This probably means modifying pull requests or using something other than a single CHANGES file.
- Once that's done, turn on protected branches

That's my thinking anyway. Protected branches make like a _lot_ easier for us, and we really should use them.

Comment (tk0miya) [reply utils/release-checklist:31]: > If we put it here, all maintainers know that we consider this part of the process to be suboptimal and that we want to improve it.

Who is "all maintainers"? This file is referred only on release operation. So it will be sawn by only me.

>That's my thinking anyway. Protected branches make like a lot easier for us, and we really should use them.

easier? really? I don't think so. It restricts developer's operation to keep repository safe. I understand it is meaningful, but it does not make our work "easier". Is my understanding wrong?

In this case, it forbids to push any commits to master branch even if it is fixing typos or update CHANGES. Is it really helpful?
If we are in large team, I agree it is very useful and we should use it ASAP. But our team is very small and inactive. Usually only I and jfbu are working.

Please let me know the advantage of protected branch **for this project**.

Note: force-pushing is already disabled. But other protection like "require status check" is not enabled.

Comment (stephenfin) [reply utils/release-checklist:31]: Before I address anything, I should point out that I'm sensing a little anger here. If so, I apologise for anything I might have said that has offended. To be clear, I'm hoping only to state my opinion and see if you agree or disagree. If you think I'm being rude, please let me know.

---

In the OpenStack world, we submit _everything_ though our code review tool, Gerrit, whether it's a huge new feature or a one word typo fix, or whether the project has hundreds of contributors or a handful. Does this introduce some overhead? Yes, of course. However, there are great advantages to this:

- Better visibility. Everyone can see the changes going in, which helps them feel more involved
- Better quality. Because everyone can see the changes going in, they have a chance to review those changes. In addition, the CI runs on everything, preventing broken master. This means users can always deploy from Git if they want cutting edge (e.g. because a bugfix isn't included in a release yet).
- Consistency. At the moment, some changes (like this one) are submitted via a PR. Others are pushed directly to master. The determination for which is which is not clear.

In addition, there are multiple maintainers and, while only two are consistently committing at the moment, surely you want this to grow in the future? Having a policy like this will help massively with onboarding these future maintainers.

Speaking from personal experience, I am not (yet!) a maintainer. As such, I have to submit my changes via pull request. However, this is a good thing. This has led to some very important comments from you and others. Likewise, I hope my comments have helped improve your own patches. I have even left comments about typos and CHANGES which you have incorporated. Any overhead of submitting PRs and waiting for CI is offset by the improvements outlined above.

Comment (stephenfin) [reply utils/release-checklist:31]: Perhaps if we are able to make `CHANGES` easier to modify, we would solve one of the things that requires you to push directly to master. What do you think of using [`reno`](https://pypi.python.org/pypi/reno), @tk0miya?

Comment (tk0miya) [reply utils/release-checklist:31]: I also have to apologize you that I was a bit frustrating in that day. I feel text based communication is very difficult again. It might come from my poor English, or a lack of mutual understanding. Anyway, we have to talk more and more to understand each other. I think you're good contributor for Sphinx. I'd still like to continue to work with you.

At last, I have to say you that I never feel you're rude. I feel you're polite man :-).

---

Thank you for explanation. I understand the advantages of branch protection you said, in general. But I wonder who wants them to this project. I know our development is inconsistent. But we have not so bad visibility and quality.

TBH, I'm afraid of take overhead of my work. Now I'm going to fix bugs, investigate troubles, improve features, refactor implements, answer questions, review PRs, update docs and do any other for Sphinx. In fact, last August and November, this project was almost stopped because I was absent. So I don't like to increase costs of my work more than now.

Anyway, I should not determine such an important rules without agreement from other maintainers (I'm not a leader!). Could you propose the idea as a new issue please (like about reno)? It should not be discussed as a comment tree of this PR. I'll agree other maintainers say okay to your proposal.

Comment (tk0miya) [reply utils/release-checklist:31]: In addition, do you think this discussion is a blocker of this PR? I think it is another topic. So I'd like to merge this if possible.

Comment (stephenfin) [reply utils/release-checklist:31]: > I also have to apologize you that I was a bit frustrating in that day. I feel text based communication is very difficult again. It might come from my poor English, or a lack of mutual understanding. Anyway, we have to talk more and more to understand each other. I think you're good contributor for Sphinx. I'd still like to continue to work with you.

Good to hear 😄 It is not problem. Just making sure I'm not being too annoying, heh

> In addition, do you think this discussion is a blocker of this PR? I think it is another topic. So I'd like to merge this if possible.

Good point, it is not. Let's discuss this in another issue 😄 The rest of the change looks good to me 👍

Comment (tk0miya) [reply utils/release-checklist:31]: Thanks :-)

**Pull request merged**
